File: web/src/lib/services/searchService.ts
import { Recipe, DietaryRestriction } from '@/types/models'

export interface SearchFilters {
  query?: string
  cuisineType?: string
  mealType?: string
  difficulty?: 'easy' | 'medium' | 'hard'
  dietaryRestrictions?: DietaryRestriction[]
  maxPrepTime?: number
  maxCookTime?: number
  minRating?: number
  ingredients?: string[]
}

export function searchRecipes(recipes: Recipe[], filters: SearchFilters): Recipe[] {
  return recipes.filter(recipe => {
    if (filters.query) {
      const query = filters.query.toLowerCase()
      const matchesName = recipe.name.toLowerCase().includes(query)
      const matchesIngredients = recipe.ingredients.some(ing => 
        ing.name.toLowerCase().includes(query)
      )
      if (!matchesName && !matchesIngredients) return false
    }
    
    if (filters.cuisineType && recipe.cuisineType !== filters.cuisineType) {
      return false
    }
    
    if (filters.mealType && recipe.mealType !== filters.mealType) {
      return false
    }
    
    if (filters.difficulty && recipe.difficulty !== filters.difficulty) {
      return false
    }
    
    if (filters.dietaryRestrictions && filters.dietaryRestrictions.length > 0) {
      const hasAllRestrictions = filters.dietaryRestrictions.every(restriction =>
        recipe.dietaryRestrictions?.includes(restriction)
      )
      if (!hasAllRestrictions) return false
    }
    
    if (filters.maxPrepTime && recipe.prepTime && recipe.prepTime > filters.maxPrepTime) {
      return false
    }
    
    if (filters.maxCookTime && recipe.cookTime && recipe.cookTime > filters.maxCookTime) {
      return false
    }
    
    if (filters.minRating && recipe.rating && recipe.rating < filters.minRating) {
      return false
    }
    
    if (filters.ingredients && filters.ingredients.length > 0) {
      const hasAllIngredients = filters.ingredients.every(filterIng =>
        recipe.ingredients.some(recipeIng => 
          recipeIng.name.toLowerCase().includes(filterIng.toLowerCase())
        )
      )
      if (!hasAllIngredients) return false
    }
    
    return true
  })
}

File: ios/Zest/Services/SearchService.swift
import Foundation

struct SearchFilters {
    var query: String?
    var cuisineType: String?
    var mealType: String?
    var difficulty: String?
    var dietaryRestrictions: [String] = []
    var maxPrepTime: Int?
    var maxCookTime: Int?
    var minRating: Double?
}

class SearchService {
    static let shared = SearchService()
    
    func search(recipes: [Recipe], filters: SearchFilters) -> [Recipe] {
        return recipes.filter { recipe in
            // Query filter
            if let query = filters.query?.lowercased(), !query.isEmpty {
                let matchesName = recipe.name.lowercased().contains(query)
                let matchesIngredients = recipe.ingredients.contains { 
                    $0.name.lowercased().contains(query) 
                }
                if !matchesName && !matchesIngredients {
                    return false
                }
            }
            
            // Cuisine filter
            if let cuisineType = filters.cuisineType, recipe.cuisineType != cuisineType {
                return false
            }
            
            // Meal type filter
            if let mealType = filters.mealType, recipe.mealType != mealType {
                return false
            }
            
            // Difficulty filter
            if let difficulty = filters.difficulty, recipe.difficulty != difficulty {
                return false
            }
            
            // Dietary restrictions filter
            if !filters.dietaryRestrictions.isEmpty {
                let hasAll = filters.dietaryRestrictions.allSatisfy { restriction in
                    recipe.dietaryRestrictions?.contains(restriction) ?? false
                }
                if !hasAll {
                    return false
                }
            }
            
            // Time filters
            if let maxPrepTime = filters.maxPrepTime, let prepTime = recipe.prepTime, prepTime > maxPrepTime {
                return false
            }
            
            if let maxCookTime = filters.maxCookTime, let cookTime = recipe.cookTime, cookTime > maxCookTime {
                return false
            }
            
            // Rating filter
            if let minRating = filters.minRating, let rating = recipe.rating, rating < minRating {
                return false
            }
            
            return true
        }
    }
}
